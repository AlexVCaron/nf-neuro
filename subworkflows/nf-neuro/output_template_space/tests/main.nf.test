nextflow_workflow {

    name "Test Subworkflow OUTPUT_TEMPLATE_SPACE"
    script "../main.nf"
    workflow "OUTPUT_TEMPLATE_SPACE"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/output_template_space"

    tag "registration"
    tag "registration/ants"
    tag "registration/antsapplytransforms"
    tag "registration/tractogram"
    tag "image"
    tag "image/applymask"
    tag "betcrop"
    tag "betcrop/fslbetcrop"
    tag "utils"
    tag "utils/templateflow"
    tag "load_test_data"

    tag "stub"
    options "-stub-run"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "tractometry.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("output to template MNI152NLin2009cAsym with local folder") {
        config "./nextflow_local.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        file("\${test_data_directory}/mni_masked.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.nii.gz")
                        ]
                    ]
                }
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM_labels_map.nii.gz")
                        ]
                    ]
                }
                input[4] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.trk"),
                            file("\${test_data_directory}/IFGWM_color.trk"),
                            file("\${test_data_directory}/IFGWM_uni.trk")
                        ]
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    ["ch_t1w_tpl", "ch_t2w_tpl"].contains(channel.key)
                                        ? file(subject).name
                                        : [ subject[0] ] + subject[1..-1].flatten().collect{ entry -> entry
                                            ? file(entry).name
                                            : "" }
                            } ]
                        }
                ).match()},
                { assert workflow.out
                    .findAll{ channel -> !channel.key.isInteger() }
                    .every{ channel -> ["ch_registered_anat",
                                        "ch_t1w_tpl",
                                        "ch_t2w_tpl",
                                        "ch_warped_labels_files",
                                        "ch_warped_nifti_files",
                                        "ch_warped_trk_files",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }

    test("output to template MNI152NLin2009aAsym - without brain mask") {
        config "./nextflow.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        file("\${test_data_directory}/mni_masked.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.nii.gz")
                        ]
                    ]
                }
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM_labels_map.nii.gz")
                        ]
                    ]
                }
                input[4] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.trk"),
                            file("\${test_data_directory}/IFGWM_color.trk"),
                            file("\${test_data_directory}/IFGWM_uni.trk")
                        ]
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    ["ch_t1w_tpl", "ch_t2w_tpl"].contains(channel.key)
                                        ? file(subject).name
                                        : [ subject[0] ] + subject[1..-1].flatten().collect{ entry -> entry
                                            ? file(entry).name
                                            : "" }
                            } ]
                        }
                ).match()},
                { assert workflow.out
                    .findAll{ channel -> !channel.key.isInteger() }
                    .every{ channel -> ["ch_registered_anat",
                                        "ch_t1w_tpl",
                                        "ch_t2w_tpl",
                                        "ch_warped_labels_files",
                                        "ch_warped_nifti_files",
                                        "ch_warped_trk_files",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }
}
