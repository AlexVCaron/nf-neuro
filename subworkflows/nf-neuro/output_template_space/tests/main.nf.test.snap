{
    "output to template MNI152NLin2009aAsym - without brain mask": {
        "content": [
            {
                "ch_registered_anat": [
                    [
                        {
                            "id": "test"
                        },
                        "test__t1_warped.nii.gz"
                    ]
                ],
                "ch_t1w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "ch_t2w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "ch_warped_labels_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM_labels_map__warped.nii.gz"
                    ]
                ],
                "ch_warped_nifti_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM__warped.nii.gz"
                    ]
                ],
                "ch_warped_trk_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM.trk",
                        "test__IFGWM_color.trk",
                        "test__IFGWM_uni.trk"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,04f69537f2d6cdc20e0a8c110ae9b7e7",
                    "versions.yml:md5,08c023826481c9130c3914096cda7f63",
                    "versions.yml:md5,0d37dad46b915d0f062b1e1c791f47a7",
                    "versions.yml:md5,a166b69e9dc7963b29c7a5944fbbf72a",
                    "versions.yml:md5,a6b014172fd4dfe073bb793b58943c46",
                    "versions.yml:md5,ad180f312de6e9124ce2789c69d3afdf",
                    "versions.yml:md5,d4da7406446940901e5757ec63bf3d8f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.5"
        },
        "timestamp": "2025-06-13T00:51:39.775526078"
    },
    "output to template MNI152NLin2009cAsym with local folder": {
        "content": [
            {
                "ch_registered_anat": [
                    [
                        {
                            "id": "test"
                        },
                        "test__t1_warped.nii.gz"
                    ]
                ],
                "ch_t1w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "ch_t2w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "ch_warped_labels_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM_labels_map__warped.nii.gz"
                    ]
                ],
                "ch_warped_nifti_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM__warped.nii.gz"
                    ]
                ],
                "ch_warped_trk_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test__IFGWM.trk",
                        "test__IFGWM_color.trk",
                        "test__IFGWM_uni.trk"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,04f69537f2d6cdc20e0a8c110ae9b7e7",
                    "versions.yml:md5,08c023826481c9130c3914096cda7f63",
                    "versions.yml:md5,0d37dad46b915d0f062b1e1c791f47a7",
                    "versions.yml:md5,a166b69e9dc7963b29c7a5944fbbf72a",
                    "versions.yml:md5,a6b014172fd4dfe073bb793b58943c46",
                    "versions.yml:md5,ad180f312de6e9124ce2789c69d3afdf",
                    "versions.yml:md5,d4da7406446940901e5757ec63bf3d8f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.5"
        },
        "timestamp": "2025-06-13T00:54:54.927818758"
    }
}